### DISCUSSION TITLE
vpcPlot, scale of X-axis, Y-axis

### DISCUSSION THREAD

**Discussion by Percival24** [nlmixr2/nlmixr2#138] [Q&A]
Hi @mattfidler, 
I was wondering if there is a way to control the number of figures that vpcPlot produces per window.  In fact, I would like especially a different scale per figure. I have 8 figures cause I have 4 studies and each study has a figure for the metabolite and the parent. The problem so far is the fact that it chooses the scale of the X-axis (resp. Y-axis) using the longest (resp. highest) time period (resp. concentration). But these scales don't suit all the figures...   
Best regards, 
Percival.

**Comment by mattfidler** [comment]
Since this is a ggplot it could probably be added, though it is outside of these packages and is found in the vpc package.  I didn't see the feature in the current `vpc` package, though it may be added (or you could try to add it yourself)

As an alternative you could use tinyvpc and see if it is implemented there. You would have to run `vpcSim()` manually for that.

A last option is to use `vpcSim()` then subset to parent/metabolite before sending to `vpc::vpc()`

**Reply by Percival24** [reply]
Here is a toy example on theophylline data reflecting exactly the problem.
First I use vpcPlot on all the fit.
Then, I try to make a subfit to have better scale for patients with stud=1, but I get stuck...

`library(ggplot2)
library(nlmixr2)
library("haven")
library("stringr")
library("ggplot2")

library(writexl)

library(xpose.nlmixr2)

require("dplyr")

library(nlmixr2)

dat1 = theo_sd
dat1$stud = 1

dat2 = theo_sd
dat2$stud = 2
dat2$ID = dat2$ID + max(dat1$ID)
dat2$DV = dat2$DV*10

dat = rbind(dat1,dat2)

one.compartment <- function() {
ini({
tka <- 0.45 # Log Ka
tcl <- 1 # Log Cl
tv <- 3.45 # Log V
b1 <- 0
b2 <- 0
b3 <- 0
eta.ka ~ 0.6
[eta.cl](http://eta.cl/) ~ 0.3
eta.v ~ 0.1
[add.sd](http://add.sd/) <- 0.7
})
model({
ka <- exp(tka + b1stud +eta.ka)
cl <- exp(tcl +b2stud +[eta.cl](http://eta.cl/))
v <- exp(tv +b3*stud +eta.v)
d/dt(depot) = -ka * depot
d/dt(center) = ka * depot - cl / v * center
cp = center / v
cp ~ add([add.sd](http://add.sd/))
})
}

fit <- nlmixr2(one.compartment, dat, est="saem", saemControl(print=0))


vpcPlot(fit,getData(fit),stratify =c("stud"))

Simu = vpcSim(fit)
Simu = Simu[Simu$stud == 1,]
vpcPlot(Simu)`

**Comment by mattfidler** [comment]
``` r
library(nlmixr2)
#> Loading required package: nlmixr2data
library(ggplot2)

dat1 = theo_sd
dat1$stud = 1

dat2 = theo_sd
dat2$stud = 2
dat2$ID = dat2$ID + max(dat1$ID)
dat2$DV = dat2$DV*1000

dat = rbind(dat1,dat2)

dat$CMT <- ifelse(dat$CMT==1, 1, dat$stud+2)


one.compartment <- function() {
  ini({
    tka <- 0.45 # Log Ka
    tcl <- 1 # Log Cl
    tv <- 3.45 # Log V
    b1 <- 0
    b2 <- 0
    b3 <- 0
    eta.ka ~ 0.6
    eta.cl ~ 0.3
    eta.v ~ 0.1
    add.sd1 <- 0.7
    add.sd2 <- 0.7
  })
  model({
    ka <- exp(tka + b1*stud + eta.ka)
    cl <- exp(tcl + b2*stud + eta.cl)
    v <- exp(tv + b3*stud + eta.v)
    d/dt(depot) = -ka * depot
    d/dt(center) = ka * depot - cl / v * center
    cp = center / v
    cp2 = cp
    cp ~ add(add.sd1)
    cp2 ~ add(add.sd2)
  })
}

fit <- nlmixr2(one.compartment, dat, est="saem", saemControl(print=0))
#> ℹ parameter labels from comments will be replaced by 'label()'
#> → loading into symengine environment...
#> → pruning branches (`if`/`else`) of saem model...
#> ✔ done
#> → finding duplicate expressions in saem model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in saem model...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> ✔ done
#> rxode2 2.0.12.9000 using 8 threads (see ?getRxThreads)
#>   no cache: create with `rxCreateCache()`
#> Calculating covariance matrix
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → loading into symengine environment...
#> → pruning branches (`if`/`else`) of saem model...
#> ✔ done
#> → finding duplicate expressions in saem predOnly model 0...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in saem predOnly model 0...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in saem predOnly model 1...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in saem predOnly model 1...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → finding duplicate expressions in saem predOnly model 2...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> → optimizing duplicate expressions in saem predOnly model 2...
#> [====|====|====|====|====|====|====|====|====|====] 0:00:00
#> ✔ done
#> → Calculating residuals/tables
#> ✔ done
#> → compress origData in nlmixr2 object, save 15296
#> → compress phiM in nlmixr2 object, save 145632
#> → compress parHist in nlmixr2 object, save 14424
#> → compress saem0 in nlmixr2 object, save 27936

Simu <- vpcSim(fit)

Simu1 <- Simu[Simu$stud == 1,]

p1 <- vpc::vpc(sim = Simu1, sim_cols = list(id = "id", dv = "sim", idv = "time"), 
    obs = dat[dat$stud == 1,], obs_cols = list(id = "ID", dv = "DV", idv = "TIME"), 
    show = list(obs_dv=TRUE)) +
  xgxr::xgx_scale_x_time_units("h") +
  ylab("Concentration Study 1") +
  rxode2::rxTheme() +
  ggplot2::ggtitle("Study 1 VPC")

Simu2 <-  Simu[Simu$stud == 2,]


p2 <- vpc::vpc(sim = Simu2, sim_cols = list(id = "id", dv = "sim", idv = "time"), 
    obs = dat[dat$stud == 2,], obs_cols = list(id = "ID", dv = "DV", idv = "TIME"), 
    show = list(obs_dv=TRUE)) +
  xgxr::xgx_scale_x_time_units("h") +
  ylab("Concentration Study 2") +
  rxode2::rxTheme() +
  ggplot2::ggtitle("Study 2 VPC")

library(patchwork)

plot(p1/p2)
```

![](https://i.imgur.com/OfaD3lC.png)

<sup>Created on 2023-04-13 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by Percival24** [reply]
When I try on my own data with another model, I get stuck with this : 

```
 Simu1 <- Simu[Simu$stud == 1,]
 p1 <- vpc::vpc(sim = Simu1, sim_cols = list(id = "id", dv = "sim", idv = "time"),
+                obs = pool[pool$stud == 1,], obs_cols = list(id = "ID", dv = "DV", idv = "TIME"),
+                show = list(obs_dv=TRUE)) +
+   xgxr::xgx_scale_x_time_units("h") +
+   ylab("Concentration Study 1") +
+   rxode2::rxTheme() +
+   ggplot2::ggtitle("Study 1 VPC")
Error in `dplyr::summarise()`:
ℹ In argument: `obs5 = quantile(dv, pi[1])`.
ℹ In group 6: `strat = 1`, `bin = 6`.
Caused by error in `quantile.default()`:
! missing values and NaN's not allowed if 'na.rm' is FALSE
Run `rlang::last_error()` to see where the error occurred.
```

But thanks for your example, it is already helpful : )

**Reply by mattfidler** [reply]
Try updating all the nlmixr2 components.  The new `nlmixr2est` 2.1.4 should fix this issue, though it samples with replacement for ODEs that cannot be solved.

**Reply by Percival24** [reply]
I got the same error message :|
The main differences btw your example and what I am dealing with are :  
- I have 4 studies, not 2. 
- The information "stud" doesn't appear in the model as a covariate.
- I have 3 other covariates in the model. 
- I have to deal with an extra compartment for the metabolite
- I have a lag time with random effect on it. 
- I have censored points.

**Reply by mattfidler** [reply]
You can add information to the simulation dataset that is not in the model with:

```r
SimuExpnd <- vpcSimExpand(fit, Simu, c("simDataColumn"))
```

Otherwise if I can't reproduce it, as much as I want to help I cannot
